File: solver.rb
require 'time'

$costs = {
  'AB' => 2 + 3 * 4 - 5 / 10.0 + 5**2,
  'AC' => 2**3 + 4 * 5 - 6 / 10.0 + 1,
  'BC' => 5 * 4 - 2 + 5**2 - 7,
  'BD' => 3 + 4 * 5 - 8 / 2.0 + 1,
  'CD' => 3**3 + 8 - 5 * 3 + 8
}

$room_gold = { 'A' => 40, 'B' => 50, 'C' => 75, 'D' => 100 }

def traverse_maze(route, gold)
  total_gold = gold

  puts "\nTrying route: #{route.join(' -> ')}"
  puts "Starting gold: #{total_gold}"
  sleep(0.25)

  (0...route.length - 1).each do |i|
    current_room = route[i]
    next_room = route[i + 1]
    path = [current_room, next_room].sort.join
    path_cost = $costs[path].ceil

    puts "\nMoving from #{current_room} to #{next_room}. Path cost: #{path_cost}"
    sleep(0.25)

    if total_gold < path_cost
      puts "Not enough gold to proceed!"
      sleep(0.25)
      return 0
    end

    total_gold -= path_cost
    puts "Paid #{path_cost} gold. Remaining gold: #{total_gold}"
    sleep(0.25)

    total_gold += $room_gold[next_room]
    puts "Collected #{$room_gold[next_room]} gold from room #{next_room}. New total: #{total_gold}"
    sleep(0.25)
  end

  total_gold
end

def find_optimal_route
  possible_routes = [
    ['A', 'B', 'D'],
    ['A', 'C', 'D'],
    ['A', 'B', 'C', 'D'],
    ['A', 'C', 'B', 'D']
  ]

  max_gold = 0
  optimal_route = []

  possible_routes.each do |route|
    gold = traverse_maze(route, 40)
    puts "\nRoute #{route.join(' -> ')} finished with #{gold} gold"
    sleep(0.25)
    if gold > max_gold
      max_gold = gold
      optimal_route = route
    end
  end

  puts "\nOptimal route: #{optimal_route.join}"
  puts "Maximum gold: #{max_gold}"
end

find_optimal_route

File: solver.js
const costs = {
    'AB': 2 + 3 * 4 - 5 / 10 + 5**2,
    'AC': 2**3 + 4 * 5 - 6 /10 + 1,
    'BC': 5 * 4 - 2 + 5**2 - 7,
    'BD': 3 + 4 * 5 - 8 / 2 + 1,
    'CD': 3**3 + 8 - 5 * 3 + 8
};

const roomGold = { 'A': 40, 'B': 50, 'C': 75, 'D': 100 };

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function traverseMaze(route, gold) {
    let totalGold = gold;

    console.log(`\nTrying route: ${route.join(' -> ')}`);
    console.log(`Starting gold: ${totalGold}`);
    await sleep(250);

    for (let i = 0; i < route.length - 1; i++) {
        const currentRoom = route[i];
        const nextRoom = route[i + 1];
        const path = [currentRoom, nextRoom].sort().join('');
        const pathCost = Math.ceil(costs[path]);

        console.log(`\nMoving from ${currentRoom} to ${nextRoom}. Path cost: ${pathCost}`);
        await sleep(250);

        if (totalGold < pathCost) {
            console.log("Not enough gold to proceed!");
            await sleep(250);
            return 0;
        }

        totalGold -= pathCost;
        console.log(`Paid ${pathCost} gold. Remaining gold: ${totalGold}`);
        await sleep(250);

        totalGold += roomGold[nextRoom];
        console.log(`Collected ${roomGold[nextRoom]} gold from room ${nextRoom}. New total: ${totalGold}`);
        await sleep(250);
    }

    return totalGold;
}

async function findOptimalRoute() {
    const possibleRoutes = [
        ['A', 'B', 'D'],
        ['A', 'C', 'D'],
        ['A', 'B', 'C', 'D'],
        ['A', 'C', 'B', 'D']
    ];

    let maxGold = 0;
    let optimalRoute = [];

    for (const route of possibleRoutes) {
        const gold = await traverseMaze(route, 40);
        console.log(`\nRoute ${route.join(' -> ')} finished with ${gold} gold`);
        await sleep(250);
        if (gold > maxGold) {
            maxGold = gold;
            optimalRoute = route;
        }
    }

    console.log(`\nOptimal route: ${optimalRoute.join('')}`);
    console.log(`Maximum gold: ${maxGold}`);
}

findOptimalRoute();